### ISSUE TITLE
Remove mutable random number generator in Gen

### ISSUE THREAD

**Issue by rickynils** [rickynils/scalacheck#67]
`Gen.Parameter` includes `rng` which is an instance of `java.util.Random`. This instance is used in `Gen.choose`, by calls to `rng.nextLong` and `rng.nextDouble`. These calls mutate the rng instance, by updating its seed. It would be nice to switch over to an immutable rng, since that would allow control over the seed (which would enable repeatable tests in ScalaCheck). So, instead of `java.util.Random` we will have `seed: Long` and `rng: Long => Long` in `Gen.Parameters`.

The change is not very big, we just need to make sure that the generators update the seed correctly. I have started to implement this, but I don't want to push the change to 1.11.0 since I've already done considerable changes there.

**Comment by non** [comment]
This was fixed in #171.

**Comment by sirthias** [comment]
My upgrade from 1.12.5 to 1.13 is currently blocked by the removal of the `rng` member on `Parameters` which I've been using to hack myself into being able to re-run tests with identical seeds.
With the 1.13 API I currently don't see how "repeatable" tests are meant to work, which were mentioned as a goal in this ticket description.

How can I manually supply the RNG seed when running a test?

**Comment by non** [comment]
@sirthias I'll look into this for you -- it may require an addition to the API.

**Comment by sirthias** [comment]
Thanks, Erik!

**Comment by non** [comment]
@sirthias If `Gen.Parameters` had:

```scala
val initialSeed: Option[Seed]
def withInitialSeed(o: Option[Seed]): Parameters
```

...and it used those as the "initial seed" when set, do you think that would be sufficient for your purposes?

**Comment by sirthias** [comment]
Thanks, Erik, for this suggestion (and sorry for my vacation-related delay in responding)!
I think this API addition would totally suffice for my purposes.

However, why would the `initialSeed` need to be an `Option`. Wouldn't it be cleaner if the `initialSeed` would always be (randomly) set right after creation and I, as a user, would simply have the option of setting it to a known value before kicking things into motion?
Maybe the difference between your proposal and my thinking is a "lazy vs. eager" decision...

**Comment by non** [comment]
@sirthias I'll push a branch with this design (and some tests proving the initial seed is working) so we can comment on something concrete.

I think this design is a bit easier to use (at least internally) because it's important that the *initial seed* be used (if present at the top-level) but ignored otherwise. If the value was not of type `Option` then I think I'd need some other mechanism for signaling whether evaluating the generator was occurring at the top-level, or from within another generator (in which case the `seed` should be threaded through, rather than initialized).

Does that make sense?

**Comment by sirthias** [comment]
Ok, I see.
For sure I don't know my way around the scalacheck internals as well as you do, so I completely trust your instincts here.
Thanks again for picking this up!

**Comment by non** [comment]
Hey @sirthias, I just pushed: https://github.com/rickynils/scalacheck/pull/263.

Take a look and let me know if you think it will work for you. If not I'm happy to make modifications as needed.
